File: main/java/uk/co/real_logic/sbe/ir/IntermediateRepresentation.java
/*
 * Copyright 2013 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.ir;

import uk.co.real_logic.sbe.util.Verify;

import java.util.*;

/**
 * Intermediate representation of SBE messages to be used for the generation of encoders and decoders
 * as stubs in various languages.
 */
public class IntermediateRepresentation
{
    private final String packageName;
    private final MessageHeader messageHeader;
    private final Map<Long, List<Token>> messagesByIdMap = new HashMap<>();
    private final Map<String, List<Token>> typesByNameMap = new HashMap<>();
    private final int version;

    /**
     * Create a new IR container taking a defensive copy of the header {@link Token}s passed.
     *
     * @param packageName that should be applied to generated code.
     * @param headerTokens representing the message header.
     */
    public IntermediateRepresentation(final String packageName, final List<Token> headerTokens, final int version)
    {
        Verify.notNull(packageName, "packageName");
        Verify.notNull(headerTokens, "headerTokens");

        this.packageName = packageName;
        this.messageHeader = new MessageHeader(Collections.unmodifiableList(new ArrayList<>(headerTokens)));
        this.version = version;
    }

    /**
     * Return the {@link MessageHeader} description for all messages.
     *
     * @return the {@link MessageHeader} description for all messages.
     */
    public MessageHeader messageHeader()
    {
        return messageHeader;
    }

    /**
     * Add a List of {@link Token}s for a given message id.
     *
     * @param messageId to identify the list of tokens for the message.
     * @param messageTokens the List of {@link Token}s representing the message.
     */
    public void addMessage(final long messageId, final List<Token> messageTokens)
    {
        Verify.notNull(messageTokens, "messageTokens");

        captureTypes(messageTokens);

        messagesByIdMap.put(Long.valueOf(messageId), Collections.unmodifiableList(new ArrayList<>(messageTokens)));
    }

    /**
     * Get the getMessage for a given identifier.
     *
     * @param messageId to get.
     * @return the List of {@link Token}s representing the message or null if the id is not found.
     */
    public List<Token> getMessage(final long messageId)
    {
        return messagesByIdMap.get(Long.valueOf(messageId));
    }

    /**
     * Get the type representation for a given type name.
     *
     * @param name of type to get.
     * @return the List of {@link Token}s representing the type or null if the name is not found.
     */
    public List<Token> getType(final String name)
    {
        return typesByNameMap.get(name);
    }

    /**
     * Get the {@link Collection} of types in for this schema.
     *
     * @return the {@link Collection} of types in for this schema.
     */
    public Collection<List<Token>> types()
    {
        return typesByNameMap.values();
    }

    /**
     * The {@link Collection} of messages in this schema.
     *
     * @return the {@link Collection} of messages in this schema.
     */
    public Collection<List<Token>> messages()
    {
        return messagesByIdMap.values();
    }

    /**
     * Get the package name to be used for generated code.
     *
     * @return the package name to be used for generated code.
     */
    public String packageName()
    {
        return packageName;
    }

    /**
     * Get the namespace name to be used for generated code.
     *
     * @return the namespace name to be used for generated code.
     */
    public String namespaceName()
    {
        return packageName;
    }

    /**
     * Get the version of the schema.
     *
     * @return version number.
     */
    public int version()
    {
        return version;
    }

    private void captureTypes(final List<Token> tokens)
    {
        for (int i = 0, size = tokens.size(); i < size; i++)
        {
            switch (tokens.get(i).signal())
            {
                case BEGIN_COMPOSITE:
                    i = captureType(tokens, i, Signal.END_COMPOSITE);
                    break;

                case BEGIN_ENUM:
                    i = captureType(tokens, i, Signal.END_ENUM);
                    break;

                case BEGIN_SET:
                    i = captureType(tokens, i, Signal.END_SET);
                    break;
            }
        }
    }

    private int captureType(final List<Token> tokens, int index, final Signal endSignal)
    {
        final List<Token> typeTokens = new ArrayList<>();

        Token token = tokens.get(index);
        typeTokens.add(token);
        do
        {
            token = tokens.get(++index);
            typeTokens.add(token);
        }
        while (endSignal != token.signal());

        typesByNameMap.put(tokens.get(index).name(), Collections.unmodifiableList(typeTokens));

        return index;
    }
}

File: main/csharp/ir/IntermediateRepresentation.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using Adaptive.SimpleBinaryEncoding.Util;

namespace Adaptive.SimpleBinaryEncoding.ir
{
    /// <summary>
    /// Intermediate representation of SBE messages to be used for the generation of encoders and decoders
    /// as stubs in various languages.
    /// </summary>
    public class IntermediateRepresentation
    {
        private readonly string _packageName;
        private readonly string _namespaceName;
        private readonly int _id;
        private readonly int _version;
        private readonly string _semanticVersion;
        
        private readonly HeaderStructure _headerStructure;
	    private readonly IDictionary<long, IList<Token>> _messagesByIdMap = new Dictionary<long, IList<Token>>();
	    private readonly IDictionary<string, IList<Token>> _typesByNameMap = new Dictionary<string, IList<Token>>();

        /// <summary>
        /// Create a new IR container taking a defensive copy of the headerStructure <seealso cref="Token"/>s passed.
        /// </summary>
        /// <param name="packageName"> that should be applied to generated code. </param>
        /// <param name="namespaceName"> that should be applied to generated code.</param>
        /// <param name="semanticVersion">semantic version for mapping to the application domain.</param>
        /// <param name="headerTokens"> representing the message headerStructure. </param>
        /// <param name="id"></param>
        /// <param name="version"></param>
        public IntermediateRepresentation(string packageName, 
            string namespaceName, 
            int id,
            int version,
            string semanticVersion,
            IList<Token> headerTokens)
	    {
		    Verify.NotNull(packageName, "packageName");
		    Verify.NotNull(headerTokens, "headerTokens");

		    _packageName = packageName;
            _namespaceName = namespaceName;
            _id = id;
		    _version = version;
            _semanticVersion = semanticVersion;
            _headerStructure = new HeaderStructure(new List<Token>(headerTokens));
        }

        /// <summary>
        /// Return the <seealso cref="HeaderStructure"/> description for all messages.
        /// </summary>
        /// <value>the &lt;seealso cref=&quot;HeaderStructure&quot;/&gt; description for all messages.</value>
        public HeaderStructure HeaderStructure
        {
            get { return _headerStructure; }
        }

        /// <summary>
	    /// Add a List of <seealso cref="Token"/>s for a given message id.
	    /// </summary>
	    /// <param name="messageId"> to identify the list of tokens for the message. </param>
	    /// <param name="messageTokens"> the List of <seealso cref="Token"/>s representing the message. </param>
	    public void AddMessage(long messageId, IList<Token> messageTokens)
	    {
            Verify.NotNull(messageTokens, "messageTokens");

		    CaptureTypes(messageTokens);

		    _messagesByIdMap[messageId] = new List<Token>(messageTokens);
	    }

	    /// <summary>
	    /// Get the getMessage for a given identifier.
	    /// </summary>
	    /// <param name="messageId"> to get. </param>
	    /// <returns> the List of <seealso cref="Token"/>s representing the message or null if the id is not found. </returns>
	    public IList<Token> GetMessage(long messageId)
	    {
		    return _messagesByIdMap[messageId];
	    }

	    /// <summary>
	    /// Get the type representation for a given type name.
	    /// </summary>
	    /// <param name="name"> of type to get. </param>
	    /// <returns> the List of <seealso cref="Token"/>s representing the type or null if the name is not found. </returns>
	    public IList<Token> GetType(string name)
	    {
		    return _typesByNameMap[name];
	    }

        /// <summary>
        /// Get the <seealso cref="Collection{T}"/> of types in for this schema.
        /// </summary>
        /// <value>the collection of types in for this schema.</value>
        public ICollection<IList<Token>> Types
        {
            get { return _typesByNameMap.Values; }
        }

        /// <summary>
        /// The collection of messages in this schema.
        /// </summary>
        /// <value>the collection of messages in this schema.</value>
        public ICollection<IList<Token>> Messages
        {
            get { return _messagesByIdMap.Values; }
        }

        /// <summary>
        /// Get the package name to be used for generated code.
        /// </summary>
        /// <value>the package name to be used for generated code.</value>
        public string PackageName
        {
            get { return _packageName; }
        }

        /// <summary>
        /// Get the namespace name to be used for generated code.
        /// </summary>
        /// <value>the namespace name to be used for generated code.</value>
        public string NamespaceName
        {
            get { return _namespaceName; }
        }

        /// <summary>
        /// Get the id number of the schema.
        /// </summary>
        public int Id
        {
            get { return _id; }
        }

        /// <summary>
        /// Get the namespaceName to be used for generated code.
        ///
        /// If <seealso cref="NamespaceName"/> is null then <seealso cref="PackageName"/> is used.
        /// </summary>
        /// <returns> the namespaceName to be used for generated code. </returns>
        public string ApplicableNamespace
        {
            get { return _namespaceName ?? _packageName; }
        }

        /// <summary>
        /// Get the version of the schema.
        /// </summary>
        /// <value>version number.</value>
        public int Version
        {
            get { return _version; }
        }

        /// <summary>
        /// Get the semantic version of the schema.
        /// </summary>
        public string SemanticVersion
        {
            get { return _semanticVersion; }
        }

        private void CaptureTypes(IList<Token> tokens)
	    {
		    for (int i = 0, size = tokens.Count; i < size; i++)
		    {
			    switch (tokens[i].Signal)
			    {
				    case Signal.BeginComposite:
					    i = CaptureType(tokens, i, Signal.EndComposite);
					    break;

				    case Signal.BeginEnum:
					    i = CaptureType(tokens, i, Signal.EndEnum);
					    break;

                    case Signal.BeginSet:
					    i = CaptureType(tokens, i, Signal.EndSet);
					    break;
			    }
		    }
	    }

	    private int CaptureType(IList<Token> tokens, int index, Signal endSignal)
	    {
		    IList<Token> typeTokens = new List<Token>();

		    Token token = tokens[index];
		    typeTokens.Add(token);
		    do
		    {
			    token = tokens[++index];
			    typeTokens.Add(token);
		    }
		    while (endSignal != token.Signal);

		    _typesByNameMap[tokens[index].Name] = typeTokens;

		    return index;
	    }
    }
}